cdap-api/src/main/java/io/cdap/cdap/api/service/Service.java

```diff
@@ -16,19 +16,49 @@
 
 package io.cdap.cdap.api.service;
 
+import io.cdap.cdap.api.ProgramLifecycle;
+import io.cdap.cdap.api.TxRunnable;
+import io.cdap.cdap.api.annotation.TransactionControl;
+import io.cdap.cdap.api.annotation.TransactionPolicy;
 import io.cdap.cdap.api.service.http.HttpServiceHandler;
 
 /**
  * Defines a custom user Service. Services are custom applications that run in program containers and provide
  * endpoints to serve requests.
  *
  * @param <T> type of service configurer
+ * @param <V> type of service context
  */
-public interface Service<T extends ServiceConfigurer> {
+public interface Service<T extends ServiceConfigurer, V extends ServiceContext> extends ProgramLifecycle<V> {
 
   /**
    * Configure the Service by adding {@link HttpServiceHandler}s to handle requests.
    * @param configurer to use to add handlers to the Service.
    */
   void configure(T configurer);
+
+  /**
+   * Initialize the {@code Service}.
+   *
+   * Note that unlike most program types, this method is not called within an implicit transaction,
+   * but instead it can start its own transactions using {@link ServiceContext#execute(TxRunnable)}.
+   * methods.
+   */
+  @Override
+  @TransactionPolicy(TransactionControl.EXPLICIT)
+  default void initialize(V context) throws Exception {
```

Comment:
Beware that this method can get called multiple times per run. Basically once per every new process

Reply:
why is it multiple times per run? Is it only if there's more than one instance of the service, or does the ServiceHttpServer get created multiple times? 

This type of information should be in the javadoc.

Reply:
Yea from my understanding is if there is more than one instance of service, it will get called per instance. @chtyim please correct me if I am wrong

Reply:
yes, that's the case.